lib/segment/src/vector_storage/dynamic_mmap_flags.rs

```diff
@@ -0,0 +1,207 @@
+use std::cmp::max;
+use std::fs::create_dir_all;
+use std::path::{Path, PathBuf};
+
+use bitvec::prelude::BitSlice;
+use memmap2::MmapMut;
+
+use crate::common::error_logging::LogError;
+use crate::common::mmap_ops::{
+    create_and_ensure_length, open_write_mmap, read_from_mmap, write_to_mmap,
+};
+use crate::entry::entry_point::{OperationError, OperationResult};
+use crate::vector_storage::chunked_utils::{ensure_status_file, MmapStatus};
+use crate::vector_storage::div_ceil;
+use crate::vector_storage::mmap_vectors::mmap_to_bitslice;
+
+#[cfg(test)]
+const MINIMAL_MMAP_SIZE: usize = 1024; // 1Kb
+
+#[cfg(not(test))]
+const MINIMAL_MMAP_SIZE: usize = 1024 * 1024; // 1Mb
+
+const FLAGS_FILE: &str = "flags.dat";
+
+pub struct DynamicMmapFlags {
+    status: MmapStatus,
+    _flags_mmap: Option<MmapMut>,
+    flags: Option<&'static mut BitSlice>,
+    status_mmap: MmapMut,
+    directory: PathBuf,
+}
+
+/// Based on the number of flags determines the size of the mmap file.
+fn mmap_capacity_bytes(len: usize) -> usize {
+    let number_of_bytes = div_ceil(len, 8);
+
+    max(MINIMAL_MMAP_SIZE, number_of_bytes.next_power_of_two())
+}
+
+/// Based on the current length determines how many flags can fit into the mmap file without resizing it.
+fn mmap_max_current_size(len: usize) -> usize {
+    let mmap_capacity_bytes = mmap_capacity_bytes(len);
+    mmap_capacity_bytes * 8
+}
+
+impl DynamicMmapFlags {
+    fn mmap_file_path(directory: &Path) -> PathBuf {
+        directory.join(FLAGS_FILE)
+    }
+
+    pub fn open(directory: &Path) -> OperationResult<Self> {
+        create_dir_all(directory)?;
+        let status_mmap = ensure_status_file(directory)?;
+        let len = *read_from_mmap(&status_mmap, 0);
+
+        let status = MmapStatus { len };
+
+        let mut flags = Self {
+            status,
+            _flags_mmap: None,
+            flags: None,
+            status_mmap,
+            directory: directory.to_owned(),
+        };
+
+        flags.reopen_mmap(len)?;
+
+        Ok(flags)
+    }
+
+    pub fn reopen_mmap(&mut self, length: usize) -> OperationResult<()> {
+        self.flags.take(); // Drop the bit slice
+        self._flags_mmap.take(); // Drop the mmap
+        let capacity_bytes = mmap_capacity_bytes(length);
+        let mmap_path = Self::mmap_file_path(&self.directory);
+        create_and_ensure_length(&mmap_path, capacity_bytes)?;
+        let mut flags_mmap = open_write_mmap(&mmap_path).describe("Open mmap flags for writing")?;
+        #[cfg(unix)]
```

Comment:
I suggest to use mmap advise wrapper introduced in `lib/segment/src/madvise.rs`

Reply:
And why do we need to specify advice flag instead of a global parameter of qdrant?